tests/org.eclipse.n4js.ui.tests/src/org/eclipse/n4js/tests/realworld/MultipleSingletonPluginTest.java

```diff
@@ -0,0 +1,200 @@
+ * this test.
+ */
+public class MultipleSingletonPluginTest extends AbstractBuilderParticipantTest {
+
+	@Inject
+	ISharedStateContributionRegistry sharedRegistry;
+
+	/**
+	 * Updates the known external library locations with the {@code node_modules} folder.
+	 */
+	@Before
+	public void setup() throws Exception {
+		setupExternalLibraries(true, true);
+	}
+
+	/**
+	 * Tries to make sure the external libraries are cleaned from the Xtext index, cleanup file system leftovers.
+	 */
+	@After
+	@Override
+	public void tearDown() throws Exception {
+		tearDownExternalLibraries(true);
+	}
+
+	/**
+	 * The tests first collects all injectors. The shared injector is identified using the class
+	 * {@link InjectorCollector}. Afterwards, it detects all singleton classes that are bound in N4JS related injectors.
+	 * Finally, for every such singleton class all injectors are checked whether they created an instance of the
+	 * singleton class and whether these instances are the same.
+	 */
+	@Test
+	public void identifyMultipleSingletons() throws Exception {
+		ProjectTestsUtils.importProject(new File("probands"), "ListBase");
+		IResourcesSetupUtil.waitForBuild();
+
+		Multimap<Class<?>, Injector> singletonInstances = HashMultimap.create();
+
+		Map<Injector, String> injectors = getAllInjectors();
+		for (Injector injector : injectors.keySet()) {
+			getN4JSSingletonsOfInjector(injector, singletonInstances);
+		}
+
+		String status = getMultipleSingletonStatusString(singletonInstances, injectors);
+
+		assertEquals(MultiSingletonExpectation.get(), status);
+	}
+
+	Map<Injector, String> getAllInjectors() {
+		Map<Injector, String> injectors = new HashMap<>();
+
+		injectors.putAll(InjectorCollector.getSharedInjectors());
+		injectors.put(new N4JSInjectorSupplier().get(),
+				"N4JS-Injector");
+		injectors.put(N4MFActivator.getInstance().getInjector(N4MFActivator.ORG_ECLIPSE_N4JS_N4MF_N4MF),
+				"N4MF-Injector");
+		injectors.put(RegularExpressionActivator.getInstance()
+				.getInjector(RegularExpressionActivator.ORG_ECLIPSE_N4JS_REGEX_REGULAREXPRESSION),
+				"Regex-Injector");
+		injectors.put(TypesActivator.getInstance().getInjector(TypesActivator.ORG_ECLIPSE_N4JS_TS_TYPES),
+				"Types-Injector");
+		injectors.put(TesterModule.getInjector(TesterModule.N4_TESTER_MODULE_ID),
+				"Tester-Injector");
+
+		return injectors;
+	}
+
+	private void getN4JSSingletonsOfInjector(Injector injector, Multimap<Class<?>, Injector> singletonInstances) {
+		for (Binding<?> b : injector.getAllBindings().values()) {
+			Key<?> key = b.getKey();
+			TypeLiteral<?> typeLiteral = key.getTypeLiteral();
+			Type type = typeLiteral.getType();
+			if (type instanceof Class<?>) {
+				Class<?> singletonClass = (Class<?>) type;
+				String singletonName = singletonClass.getName();
+				Singleton annotation = singletonClass.getAnnotation(Singleton.class);
+
+				if (annotation != null && singletonName.toLowerCase().contains("n4js")) {
```

Comment:
will this catch singletons that are not declared as such, but bound as such?

https://github.com/google/guice/wiki/Scopes#eager-singletons

Reply:
also this filtering by name... we are not consistent with naming, something might slip through

Reply:
1) as discussed: I'll change the `in(SINGLETON)` methods. Also I add a comment.
2) also the package name is taken into account and it should contain ".n4js."